onnxscript/ir/passes/_pass_infra.py

```diff
@@ -97,76 +146,112 @@ def ensures(self, model: ir.Model) -> None:
         del model  # Unused
 
 
-class PassManager:
+class InPlacePass(PassBase):
+    """A pass that modifies the input model in place and returns it."""
+
+    @property
+    def in_place(self) -> bool:
+        return True
+
+    @property
+    def changes_input(self) -> bool:
+        return True
+
+
+class OutOfPlacePass(PassBase):
+    """A pass that returns a new model but does not modify the input model."""
+
+    @property
+    def in_place(self) -> bool:
+        return False
+
+    @property
+    def changes_input(self) -> bool:
+        return False
+
+
+class DestructivePass(PassBase):
```

Comment:
Do we really need this? Is there any good example where this is useful? Seems to me it is better to encourage users to use just one of above two ... (which may mean we don't need these two boolean properties) ... I can see the value of "analysis" passes that compute some information but don't modify the model, but not sure whether those are within the scope of "passes" as designed ...

Reply:
https://github.com/microsoft/onnxscript/pull/2119 this is an example where the pass is DestructivePass. Maybe possible to implement it as a copy as well? I can remove the class to discourage the use though.

Reply:
Done

Reply:
> #2119 this is an example where the pass is DestructivePass. Maybe possible to implement it as a copy as well? I can remove the class to discourage the use though.
> 
> (but the boolean properties are still useful. For example, we can turn a destructive pass to an out of place pass by chaining a "Copy" pass with the destructive pass automatically by looking at these two properties)

I see. I think I would classify #2219 as an example of an in-place pass that returns some extra information (the extracted sub-model) ... and a version that returns only the extracted sub-model without modifying the input-model could be an out-of-place pass (though, in general, we might need the ability to return values other than a model for more complex analysis passes).

Reply:
Although: the two properties are still useful I suppose: When we compose passes with into a sequence, whether a pass is out of place or destructive requires some careful logic: https://github.com/microsoft/onnxscript/pull/2120/files#diff-70c7e5b3422f4daaf1611d4f76578c96e4c5894cced3d51718efa0290219f7f5R180-R186